lib/Sema/TypeCheckAttr.cpp

```diff
@@ -1732,7 +1762,8 @@ void AttributeChecker::visitSpecializeAttr(SpecializeAttr *attr) {
   }
 
   // Check the validity of provided requirements.
-  checkSpecializeAttrRequirements(attr, FD, resolvedRequirements, TC);
+  checkSpecializeAttrRequirements(attr, FD, resolvedRequirements,
+                                  constrainedGenericParams, TC);
 
   // Store converted requirements in the attribute so that they are
```

Comment:
This code goes through a lot of trouble to build the convertedRequirements. However much of the logic could go away if we just used the generic signature from the Builder instead. It seems we're only using the ArchetypeBuilder for diagnostics, and then duplicating a lot of work to canonicalize the requirements.